content/compute-app/non-functional-requirements/readme.md

```diff
@@ -0,0 +1,156 @@
+   - In a hot-cold architecture, you might have last replicated or backed up your data the previous night.
+   - In a warm-standby architecture, you might have asynchronous replication of your data across regions, and maybe it's ~15 minutes out of date.
+   - This is your recovery point.
+- **RPO** is the amoiunt of data that you can tolerate losing in the event of a failure.
+
+- Think about the business continuity in a wider context - e.g. can you switch to manual operations?
+
+- Azure services provide several options for how you architect for different levels of uptime, RTO, and RPO. But these have tradeoffs.
+   - Single-region, single-zone architectures tend to be cheapest, but have the least resiliency. You likely can't support low RTO or RPO.
+   - Multi-region designs are often expensive because you need to deploy multiple separate instances. They can also introduce complexity. And data replication is usually asynchronous, which means you don't automatically get low RPO. But your risk is spread across multiple regions, so they can be resilient and performant, especially when you have customers in multiple regions.
+   - Single-region, multi-zone architectures tend to be the sweet spot for most customers. There is still some cost involved in the redundancy, but you can often support low RPO and RTO requirements.
+  - Also, you might choose to have different controls in different parts of your system. For example, if you ingest invoices and then process them:
+    - You might decide to synchronously replicate the incoming invoice data - it's hard to replace this, and even though there's a perf hit on the synchronous replication, it's worth it.
+    - But then the other data (e.g. everything derived in processing the invoice) isn't synchronously replicated, because you can re-process that if you really have to, and so you can prioritize cost and performance for this set of data.
+    - So you need to understand your business and business requirements to sensibly decide on these approaches - which is why these requirements and decisions should be led by informed business stakeholders, not by the technology teams. 
+
+- Some other things we want to touch on around tradeoffs of resiliency vs. other requirements:
+  - What happens when customer tells they need 100% availability?
+  - Design complexity comparison wrt SLAs (three 9's vs 4 9's for example)
+  - What happens when customer tells they need 100% recoverability (RPO & RTO as zero)
+  - Impact of recoverability metrics in BCDR design
+
+## Performance
+
+- **Pillar Concept** - performance baseline
+
+- Need to think about overall performance as well as performance of individual components.
+
+- For example, imagine you're building a solution that handles processing of invoices. You might care about both:
+   - End-to-end transaction latency
+   - The response time of individual APIs
+- Make sure you're clear about whether performance requirements are hard or soft (SLAs or SLOs).
```

Comment:
Maybe some high-level guidance on how to decide on what is reasonable performance?

Reply:
Good idea. We've just thought about this and came up with a few questions for customers to think through when deciding this. It's hard to give clear metric-driven guidance, but this might help to narrow down to an order of magnitude:

- What's the system all about? Is it a business-critical or low-priority system? Is it used a lot or infrequently?
- What will the users generally notice? When would they start to complain?
- Are you over-optimising a single part of a wider end-to-end system?
- What's reasonable given your constraints? e.g. if you rely on a database or a downstream third-party system, you might not have much ability to influence performance.
- If you're migrating or replacing an existing system, have you got a baseline measurement of performance? Is this acceptable or too slow?

@arsenvlad any thoughts on further questions we should be posing?

Reply:
The questions above are awesome. Maybe just concretizing it via an example like a web application where some latency of 100ms maybe OK and trying to get it to <10ms maybe too much... versus a high frequency trading platform where every ms matters. Another point to mention about performance is that relying just on average is not enough, often need to look at percentiles to set expectations around outliers.